packages/paste-core/components/popover/src/PopoverContainer.tsx

```diff
@@ -1,17 +1,35 @@
 import * as React from 'react';
-import {
-  useNonModalDialogPrimitiveState,
+import {Box} from '@twilio-paste/box';
+import type {
+  NonModalDialogPrimitiveStateReturn,
   NonModalDialogPrimitivePopoverInitialState,
 } from '@twilio-paste/non-modal-dialog-primitive';
+import {useNonModalDialogPrimitiveState} from '@twilio-paste/non-modal-dialog-primitive';
 import {PopoverContext} from './PopoverContext';
 
-export type PopoverContainerProps = NonModalDialogPrimitivePopoverInitialState;
+export interface PopoverStateReturn extends NonModalDialogPrimitiveStateReturn {
+  [key: string]: any;
+}
 
-const PopoverContainer: React.FC<PopoverContainerProps> = ({children, placement, gutter, ...initialState}) => {
-  const popover = useNonModalDialogPrimitiveState({placement, gutter, modal: true, ...initialState});
-  const value = React.useMemo(() => popover, Object.values(popover));
-  return <PopoverContext.Provider value={value}>{children}</PopoverContext.Provider>;
-};
+export interface PopoverContainerProps extends NonModalDialogPrimitivePopoverInitialState {
+  children: NonNullable<React.ReactNode>;
+  state?: PopoverStateReturn;
+}
+
+const BasePopoverContainer = React.forwardRef<HTMLDivElement, PopoverContainerProps>(
+  ({children, gutter, placement, state, ...initialState}, ref) => {
+    const popover = state || useNonModalDialogPrimitiveState({gutter, modal: true, placement, ...initialState});
+    return (
+      <PopoverContext.Provider value={popover}>
+        <Box ref={ref}>{children}</Box>
+      </PopoverContext.Provider>
+    );
+  }
+);
+
+const PopoverContainer = React.memo(BasePopoverContainer);
```

Comment:
I honestly don't know the answer to this, but does wrapping the component in memo do anything weird to the ref, like not forward it? Does it still work?

Reply:
Did some searching and didn't find anything saying NOT to do wrap components that use `forwardRef` in `memo()`. I also tested locally, and the `memo` function seems to have no effect on the `ref`. So I think we're good to go.

Reply:
👍 Bonza